pkg/controller/service/service_controller.go

```diff
@@ -446,13 +328,17 @@ func (s *ServiceController) needsUpdate(oldService *v1.Service, newService *v1.S
 			oldService.Spec.Type, newService.Spec.Type)
 		return true
 	}
-
+	if !hasFinalizer(newService) && newService.DeletionTimestamp == nil {
+		// We only need to update the service if it's missing the finalizer
+		// and not marked for deletion. If it is marked and missing that means it
+		// was successfully removed already.
+		return true
```

Comment:
Trying to understand the reasoning --- under what condition we will have a load balancer service has no finalizer and not being deleted?

Reply:
All LB services that existed prior to the upgrade. This ensures the initial sync of the controller will add the finalizers to existing services.